sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -510,17 +510,6 @@ class Analyzer(
           ordering.map(order => resolveExpression(order, child).asInstanceOf[SortOrder])
         Sort(newOrdering, global, child)
 
-      // A special case for Generate, because the output of Generate should not be resolved by
-      // ResolveReferences. Attributes in the output will be resolved by ResolveGenerate.
-      case g @ Generate(generator, join, outer, qualifier, output, child)
-        if child.resolved && !generator.resolved =>
```

Comment:
This trick does not always work, https://github.com/apache/spark/pull/11497#discussion_r55021069 is a good example.

However, instead of adding another trick in https://github.com/apache/spark/pull/11497, this PR tries to solve the problem fundamentally.